### ISSUE TITLE
Feature: Better error handling for ngRoute

### ISSUE THREAD

**Issue by baym0n** [angular/angular.js#13931]
For a route definition, a comfy error handling would be nice.

```
.config(function($routeProvider) {
    $routeProvider
    .when('/', {
        templateUrl: 'view/welcome.html'
    })
    .when('/page', {
        templateUrl: 'view/page.html'
    })
    .otherwise({
        templateUrl: 'view/404.html'
    });
})
```

If one or more of the templates (referenced by url) could not be loaded from the server, a fallback strategy is required. For example something like this:

```
.config(function($routeProvider) {
    $routeProvider
    .when('/', {
        templateUrl: 'view/welcome.html'
    })
    .when('/page', {
        templateUrl: 'view/page.html'
    })
    .otherwise({
        templateUrl: 'view/404.html'
    })
    .onError({
        template: 'Template could not be loaded'
    });
})
```

The onError "route" is automatically used, if there if a template could not be resolved.

EDIT: [see also here](http://stackoverflow.com/questions/35162551/angularjs-fallback-strategy-when-template-not-found), there is actually no working way I could find to do that.

**Comment by Narretz** [comment]
Under which circumstances do you expect your route templates not to be found? Are you loading arbitrary templates? Not-found should basically not happen if you put your templates in the templateCache in a build step.
Unfound templates also trigger a $routeChangeError which you can listen to. It's not the most convenient way (and you'll have an error in your console), but it's possible to do something about this.

**Comment by baym0n** [comment]
> Under which circumstances do you expect your route templates not to be found?

In my case the templates are loaded from the web server, So, network issues, server problems, IO errors, invalid configurations, firewall settings, wrong permissions, ... this seems to be the most obvious "circumstances". Assuming that there is always just the happy path is not an option. If you say failures "should [...] not happen", they WILL happen, sooner or later.

I already know about the `$routeChangeError`, see my [question](http://stackoverflow.com/questions/35162551/angularjs-fallback-strategy-when-template-not-found) at stack overflow. If you tell me how to set the template directly from the callback handler and if you are not interested into optimization, you can close this issue. But until now I could not find a comfy way to do that.

Hint: what I DON'T want:

- replacing the location path with something other or redirect to another page
- printing an error message through console.log() (this is a technical tool for development, maybe operations either, not a method to inform about a problem)
- still showing the previous view with the already updated location path (which is the current situation)

I want to show a user/customer suitable error message in a useful manner for exactly that location path which has causes the error. that means I have to set the template directly when this error occurs.

Inline template would be great for that.

Template:

    <script type="text/ng-template" id="template-error">
        ...
    </script>

Reference:

    .onError({
        templateId: 'template-error'
    });

**Comment by Narretz** [comment]
Yeah, your requirements might not be possible to fulfill with the current state.
Still, if you know what your templates are in advance, the easiest way is to put them into the $templateCache in a build step. See https://github.com/karlgoldstein/grunt-html2js for a grunt solution. That takes network loading completely out of the equation.

**Comment by baym0n** [comment]
> Yeah, your requirements might not be possible to fulfill with the current state.

Too bad, I hoped I don't have the hack this into the ngView/ngRoute modules by myself :(

**Comment by Narretz** [comment]
So what's wrong with the build step approach?

**Comment by baym0n** [comment]
This may works, but feels more like a workaround.

**Comment by Narretz** [comment]
It is a workaround, but it has the benefit, that you won't have to handle network problems for any template ever. It also eliminates the delay when the template is first loaded from the server.

**Comment by Narretz** [comment]
There's one way to do it. Use an http interceptor, and react to errors for specific url patterns: http://plnkr.co/edit/R83oHVcpWJ8oGkGCnfJp?p=preview

**Comment by baym0n** [comment]
> There's one other way to do it. Use an http interceptor, and react to errors for specific url patterns: http://plnkr.co/edit/R83oHVcpWJ8oGkGCnfJp?p=preview

Now we're talking... I tried that and it suits my purpose. You really should add this to your official documentation! ... and maybe optimize the possibilities of error handling for template loading issues ;)

Btw: the "grunt-html2js" solution may only works for smaller projects. For projects with a massive amount of views this could be problematic.

**Comment by Narretz** [comment]
Closing, because the use case is pretty specific and a solution can be achieved with a http interceptor.

**Comment by baym0n** [comment]
The problem with the interceptor is, that the template system never tries to get the real(!) template again, because the interceptor has returned a template value. So this is what it is: a workaround. Too bad that you decided to not increase the resilience of your framework - for no reason. This makes angular not a good solution for big projects and real world scenarios (as described above).

And this is not a use case, this is a failure scenario. And yes, it is specific ... and it is real.

**Comment by Narretz** [comment]
Your previous comment sounded a lot more satisfied with the solution ... sorry to hear that it's not working out as planned. 
We are aware that ngRoute is not the best routing module we could have provided for angular core. For more requirements, the ui-router has always been our recommendation.
And it's not "no reason". The reason is also that it's just not a highly demanded feature, as you can see from this issue.
One could also argue that replacing the actual template content with a placeholder is what gets you into the issue that you now never have access to the real template. I would recommend opening a modal or a "toast" with the information that the server could not be reached

Btw, The interceptor logic can also handle retries: http://plnkr.co/edit/A6kH2wJImLo95vhewV3J?p=preview 
Basically, I put every failed templateRequest inside an array, and add a request interceptor that removes the cache property from the config, so a new request is made. Yes, that's not obvious from the docs, but it's definitely a flexible solution.
